### PR TITLE
fix(grouping): Do not accept Swift attributes as function names

### PR DESCRIPTION
**Pull request by jjbayer**
For anonymous functions, our function name trimming logic would pick a swift attribute as function name, e.g. `@callee_guaranteed` for `thunk for @escaping @callee_guaranteed () -> ()`.

Set function name to `<anonymous function>` if the picked function name is an attribute.

Fix https://getsentry.atlassian.net/browse/INGEST-164

### CODE REVIEW

File: tests/sentry/stacktraces/test_functions.py [hunk 1]
```diff
@@ -109,6 +109,40 @@ def test_trim_csharp_function_name(input, output):
...
+            "thunk/partial apply for anonymous function",
+        ],
+        [
+            "partial apply for thunk for @escaping @callee_guaranteed (@guaranteed SomeType, @guaranteed [String : SomeType2], @guaranteed SomeType3) -> ()",
+            "thunk/partial apply for anonymous function",
+        ],
+        [  # "closure ... in ..." functions are converted to containing
+            # function. We might want to change this in the future
+            "closure #1 (T1) in foo(bar: T2)",
+            "foo",
```
Comment (jjbayer) [inline tests/sentry/stacktraces/test_functions.py:134]: @untitaker This stays the way it was, but it might be worth opening another PR which replaces "foo" by "closure" in these cases.

File: src/sentry/stacktraces/functions.py [hunk 2]
```diff
@@ -214,6 +215,10 @@ def process_generics(value, start):
             func_token = None
 
     if func_token:
+        if func_token.startswith("@") and platform in ("cocoa", "swift"):
+            # Found a Swift attribute instead of a function name, must be an
+            # anonymous function
+            func_token = "<anonymous function>"
```
Comment (jjbayer) [inline src/sentry/stacktraces/functions.py:221]: @jan-auer is `<anonymous function>` a good placeholder? Would something language-specific be better, e.g. "closure" (not even sure if every anon. function is called a closure)?

**Review by untitaker — APPROVED** [review]

File: src/sentry/stacktraces/functions.py (hunk 2 above)
Comment (jan-auer) [reply src/sentry/stacktraces/functions.py:220]: Let's ask @philipphofmann how he would call this. Since you're detecting the platform, something language specific would probably be best.

**Review by jan-auer — COMMENTED** [review]
In the case of `thunk for @escaping @callee_guaranteed () -> ()`, we should also name this `thunk for <anonymous function>` or something similar. It does get a bit longer, but makes it clear that this is _not_ the function itself but some autogenerated code.

File: src/sentry/stacktraces/functions.py (hunk 2 above)
Comment (philipphofmann) [reply src/sentry/stacktraces/functions.py:220]: This is a complex topic. `thunk for @escaping @callee_guaranteed () -> ()` are keywords from the Swift Intermediate Language, see https://github.com/apple/swift/blob/main/docs/SIL.rst#function-types. To give you a better answer, I would need some time for investigation. Of course, you can as well read into that topic on your own.

**Comment by jjbayer** [comment]
> In the case of `thunk for @escaping @callee_guaranteed () -> ()`, we should also name this `thunk for <anonymous function>` or something similar. It does get a bit longer, but makes it clear that this is _not_ the function itself but some autogenerated code.

@jan-auer We can do that, but note that we have more problems like this, e.g.

* `closure #1 (T1) in foo(bar: T2)` currently becomes `foo`,
* `partial apply for thunk for (...) -> (...)` -> What should this become?

File: src/sentry/stacktraces/functions.py (hunk 2 above)
Comment (philipphofmann) [reply src/sentry/stacktraces/functions.py]: @brustolin, do you maybe have input on this one?

**Review by untitaker — APPROVED** [review]

**Comment by jjbayer** [comment]
After a discussion with @jan-auer, we decided to go with `closure` / `thunk for closure`. This might not be perfect but better than the status quo. There is still future work to be done in parsing swift function names, see examples in comment above / added test cases.
